src/net/server/middleware/tsig.rs

```diff
@@ -0,0 +1,496 @@
+                // service pushes response messages into the stream and we try
+                // and sign them. If there isn't enough space to add the TSIG
+                // signature RR to the response we can't signal the upstream
+                // to try again to produce a smaller response message as it
+                // may already have finished pushing into the stream or be
+                // several messages further on in its processsing. We also
+                // can't edit the response message content ourselves as we
+                // know nothing about the content. The only option left to us
+                // is to try and truncate the TSIG MAC and see if that helps,
+                // but we don't support that (yet? NSD doesn't support it
+                // either).
+                return Some(mk_error_response(
+                    request.message(),
+                    OptRcode::SERVFAIL,
+                ));
+            }
+        }
+
+        None
+    }
+
+    fn mk_signed_truncated_response(
+        request: &Request<RequestOctets>,
+        key: &tsig::Key,
+    ) -> Option<AdditionalBuilder<StreamTarget<NextSvc::Target>>> {
+        let octets = request.message().as_slice().to_vec();
+        let mut mut_msg = Message::from_octets(octets).unwrap();
+        let res =
+            ServerTransaction::request(&key, &mut mut_msg, Time48::now());
+
+        match res {
+            Ok(None) => {
+                warn!("Ignoring attempt to create a signed truncated response for an unsigned request.");
+                None
+            }
+
+            Ok(Some(tsig)) => {
+                let builder = mk_builder_for_target();
+                let mut new_response = builder
+                    .start_answer(request.message(), Rcode::NOERROR)
+                    .unwrap();
+                new_response.header_mut().set_tc(true);
+                let mut new_response = new_response.additional();
+
+                if let Err(err) =
+                    tsig.answer(&mut new_response, Time48::now())
+                {
+                    error!("Unable to sign truncated TSIG response: {err}");
+                    Some(mk_error_response(
+                        request.message(),
+                        OptRcode::SERVFAIL,
+                    ))
+                } else {
+                    Some(new_response)
+                }
+            }
+
+            Err(err) => {
+                error!("Unable to sign truncated TSIG response: {err}");
+                Some(mk_error_response(request.message(), OptRcode::SERVFAIL))
+            }
+        }
+    }
+
+    fn get_relevant_question(
+        msg: &Message<RequestOctets>,
+    ) -> Option<Question<ParsedName<RequestOctets::Range<'_>>>> {
+        if Opcode::QUERY == msg.header().opcode() && !msg.header().qr() {
+            if let Some(q) = msg.first_question() {
+                if matches!(q.qtype(), Rtype::SOA | Rtype::AXFR | Rtype::IXFR)
```

Comment:
While not a requirement of RFC 8945, one could argue that this violates the intent of the RFC because the RFC refers to DNS UPDATE which uses the UPDATE opcode, but this requires the QUERY opcode. The thinking here was that we are adding TSIG support to authenticate zone transfers, which is the other use case described by RFC 8945. Is this too strict? If this test fails the middleware will allow the request to proceed, so this is defining the set of request types which will be TSIG validated. Perhaps we should allow all requests? (but not checking opcode could be problematic, perhaps some opcodes change the message format in ways incompatible with TSIG?). @partim, @Philip-NLnetLabs: What do you think?

Reply:
I think that for a middleware TSIG layer it is best to reject what we cannot support at the moment. So if get_relevant_question fails then don't proceed.

Reply:
Maybe a FORMERROR is most clear if we get a TSIG record with something the TSIG middleware doesn't support.

Reply:
I think it depends what you mean by "cannot support".

Currently if a TSIG signed request is received by the TSIG middleware but is an opcode or qtype that is not considered "relevant" then the TSIG middleware will ignore it and pass it on unchanged to the next layer, either another middleware service or the application service. The TSIG RR would be left intact in the additional section. 

Middleware or application service layers that really need to know whether or not the request was considered validly signed could check the `Option<Key>` passed out by the TSIG middleware service to check (a) is it validly signed (yes if `Some`, unsigned if `None`) and (b) if signed, was it signed with the required key?

"Irrelevant" queries have to be passed on unchanged otherwise requests of other types (e.g. a query for an AAAA record) would not propagate through the service layers even not signed.

My question was actually about something else: should we check for and validate TSIG signatures on more/all types of request rather than just the limited set it checks for now?

Reply:
We can't do update at the moment. So I'd say supporting query and notify should be enough.

Reply:
My comment got posted below instead of here:

> After some internal discussion it seems to me that there is no reason to restrict the opcode or query type on which TSIG verification is performed. If the request has a TSIG RR and the server stack includes TSIG middleware then that middleware should verify the TSIG signature and also sign any corresponding response. So I'm going to remove this restriction entirely.

Regarding your response, we don't do UPDATE but we can verify TSIG on UPDATE requests as it is no different than verifying TSIG on any other request. As stated above I intend to remove the check entirely, let library users decide if they want / need more restrictions, and let it be used wherever it adds value, which is also how NSD works.